src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResultBuilder.cs

```diff
@@ -0,0 +1,359 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics.Telemetry;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Stores the results of analyzer execution:
+    /// 1. Local and non-local diagnostics, per-analyzer.
+    /// 2. Analyzer execution times, if requested.
+    /// </summary>
+    internal class AnalysisResultBuilder
+    {
+        private readonly object _gate = new object();
+        private readonly Dictionary<DiagnosticAnalyzer, TimeSpan> _analyzerExecutionTimeOpt;
+        private readonly HashSet<DiagnosticAnalyzer> _completedAnalyzers;
+        private readonly Dictionary<DiagnosticAnalyzer, AnalyzerActionCounts> _analyzerActionCounts;
+
+        private Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>.Builder>> _localSemanticDiagnosticsOpt = null;
+        private Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>.Builder>> _localSyntaxDiagnosticsOpt = null;
+        private Dictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>.Builder> _nonLocalDiagnosticsOpt = null;
+        
+        internal AnalysisResultBuilder(bool logAnalyzerExecutionTime, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            _analyzerExecutionTimeOpt = logAnalyzerExecutionTime ? CreateAnalyzerExecutionTimeMap(analyzers) : null;
+            _completedAnalyzers = new HashSet<DiagnosticAnalyzer>();
```

Comment:
> HashSet doesn't have a constructor taking capacity - leaving this as is.

It will do in the future though, although I guess that doesn't help you here? I proposed it as an [CoreFx Issue](https://github.com/dotnet/corefx/issues/382) and it got accepted, apparently it's available in the [future branch](https://github.com/dotnet/corefx/issues/382#issuecomment-132885883).